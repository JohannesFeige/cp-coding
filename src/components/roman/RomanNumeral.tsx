import React, { useState, useCallback } from 'react';
import { toRomanNumerals } from '../shared/calculation.service';

const RomanNumeral: React.FC = () => {
  const [value, setValue] = useState<undefined | number>();
  const [romanValue, setRomanValue] = useState<string>();

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = Number.parseInt(event.target.value);
    setValue(newValue);
    setRomanValue(getTranslation(newValue));
  };

  const getTranslation = useCallback((value: number | undefined) => {
    let tranlatedValue = '';
    if (value) {
      tranlatedValue = toRomanNumerals(value);
    }
    return tranlatedValue;
  }, []);

  return (
    <div>
      <div className="input-group mb-3">
        <div className="input-group-prepend">
          <span className="input-group-text">Decimal Value</span>
        </div>
        <input
          type="number"
          className="form-control"
          placeholder="Decimal Value"
          value={value}
          onChange={handleChange}
        />
      </div>
      <span id="result">Roman Numeral: {romanValue}</span>
    </div>
  );
};

export default RomanNumeral;
