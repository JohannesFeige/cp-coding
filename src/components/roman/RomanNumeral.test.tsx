import React from 'react';
import RomanNumeral from './RomanNumeral';
import { shallow, ShallowWrapper } from 'enzyme';

describe('RomanNumeral', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<RomanNumeral />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('should render', () => {
    expect(wrapper).not.toBe(null);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render and tranlate', () => {
    expect(wrapper).not.toBe(null);
    const input = wrapper.find('input[type="number"]');

    expect(input.getElements().length).toBe(1);
    input.simulate('change', {
      target: {
        value: 42,
      },
    });

    const result = wrapper.find('span');
    expect(result.getElements()).toHaveLength(2);
    expect(result.last().text()).toBe('Roman Numeral: XLII');
  });
});
