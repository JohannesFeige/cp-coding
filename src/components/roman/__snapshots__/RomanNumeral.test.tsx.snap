// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RomanNumeral should render 1`] = `
<div>
  <div
    className="input-group mb-3"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        Decimal Value
      </span>
    </div>
    <input
      className="form-control"
      onChange={[Function]}
      placeholder="Decimal Value"
      type="number"
    />
  </div>
  <span
    id="result"
  >
    Roman Numeral: 
  </span>
</div>
`;
